File: modules/tests/src/commonTest/kotlin/arrays/CallbackArgArray.kt
package arrays

import natives.test.MyDictionary
import natives.test.MyEnum
import withLib
import kotlin.test.Test
import kotlin.test.assertTrue

class CallbackArgArray {

    @Test
    fun callbackArgCharArray() = withLib {
        assertTrue(natives.test.callbackArgCharArray { it.contentEquals(charArrayOf('a', 'b')) })
    }

    @Test
    fun callbackArgBooleanArray() = withLib {
        assertTrue(natives.test.callbackArgBooleanArray { it.contentEquals(booleanArrayOf(true, false)) })
    }

    @Test
    fun callbackArgByteArray() = withLib {
        assertTrue(natives.test.callbackArgByteArray { it.contentEquals(byteArrayOf(1, 2)) })
    }

    @Test
    fun callbackArgShortArray() = withLib {
        assertTrue(natives.test.callbackArgShortArray { it.contentEquals(shortArrayOf(1, 2)) })
    }

    @Test
    fun callbackArgIntArray() = withLib {
        assertTrue(natives.test.callbackArgIntArray { it.contentEquals(intArrayOf(1, 2)) })
    }

    @Test
    fun callbackArgLongArray() = withLib {
        assertTrue(natives.test.callbackArgLongArray { it.contentEquals(longArrayOf(1, 2)) })
    }

    @Test
    fun callbackArgFloatArray() = withLib {
        assertTrue(natives.test.callbackArgFloatArray { it.contentEquals(floatArrayOf(1.1f, 2.2f)) })
    }

    @Test
    fun callbackArgDoubleArray() = withLib {
        assertTrue(natives.test.callbackArgDoubleArray { it.contentEquals(doubleArrayOf(1.1, 2.2)) })
    }

    @Test
    fun callbackArgStringArray() = withLib {
        assertTrue(natives.test.callbackArgStringArray { it.contentEquals(arrayOf("string1", "string2")) })
    }

    @Test
    fun callbackArgEnumArray() = withLib {
        assertTrue(natives.test.callbackArgEnumArray { it.contentEquals(arrayOf(MyEnum.CASE1, MyEnum.CASE2)) })
    }

    @Test
    fun callbackArgDictionaryArray() = withLib {
        assertTrue(natives.test.callbackArgDictionaryArray {
            it.contentEquals(arrayOf(
                MyDictionary(1, 2, 3, 4),
                MyDictionary(5, 6, 7, 8)
            ))
        })
    }
}

File: modules/tests/natives/testcpp/src/arrays/callback_arg_array.cpp
#include <api.hpp>

bool callback_arg_char_array(std::shared_ptr<CallbackPassCharArray> arg) {
    return arg->invoke(KArray<uint16_t> {
        'a', 'b'
    });
}

bool callback_arg_char_array_n(std::shared_ptr<CallbackPassCharArrayN> arg) {
    return arg->invoke(KOptional<KArray<uint16_t>>());
}

bool callback_arg_boolean_array(std::shared_ptr<CallbackPassBooleanArray> arg) {
    return arg->invoke(KArray {
        true, false
    });
}

bool callback_arg_byte_array(std::shared_ptr<CallbackPassByteArray> arg) {
    return arg->invoke(KArray<int8_t> {
        1, 2
    });
}

bool callback_arg_ubyte_array(std::shared_ptr<CallbackPassUByteArray> arg) {
    return arg->invoke(KArray<uint8_t> {
        1u, 255u
    });
}

bool callback_arg_short_array(std::shared_ptr<CallbackPassShortArray> arg) {
    return arg->invoke(KArray<int16_t> {
        1, 2
    });
}

bool callback_arg_ushort_array(std::shared_ptr<CallbackPassUShortArray> arg) {
    return arg->invoke(KArray<uint16_t> {
        1u, 65535u
    });
}

bool callback_arg_int_array(std::shared_ptr<CallbackPassIntArray> arg) {
    return arg->invoke(KArray {
        1, 2
    });
}

bool callback_arg_uint_array(std::shared_ptr<CallbackPassUIntArray> arg) {
    return arg->invoke(KArray {
        1u, 4294967295u
    });
}

bool callback_arg_long_array(std::shared_ptr<CallbackPassLongArray> arg) {
    return arg->invoke(KArray<int64_t> {
        1, 2
    });
}

bool callback_arg_ulong_array(std::shared_ptr<CallbackPassULongArray> arg) {
    return arg->invoke(KArray<uint64_t> {
        1, 18446744073709551615u
    });
}

bool callback_arg_float_array(std::shared_ptr<CallbackPassFloatArray> arg) {
    return arg->invoke(KArray {
        1.1f, 2.2f
    });
}

bool callback_arg_double_array(std::shared_ptr<CallbackPassDoubleArray> arg) {
    return arg->invoke(KArray {
        1.1, 2.2
    });
}

bool callback_arg_string_array(std::shared_ptr<CallbackPassStringArray> arg) {
    return arg->invoke(KArray {
        KString("string1"),
        KString("string2")
    });
}

bool callback_arg_string_array_n(std::shared_ptr<CallbackPassStringArrayN> arg) {
    return arg->invoke(KArray {
        KOptional<KString>(),
        KOptional<KString>()
    });
}

bool callback_arg_enum_array(std::shared_ptr<CallbackPassEnumArray> arg) {
    return arg->invoke(KArray {
        CASE1, CASE2
    });
}

bool callback_arg_dictionary_array(std::shared_ptr<CallbackPassDictionaryArray> arg) {
    return arg->invoke(KArray {
        MyDictionary(1, 2, 3, 4),
        MyDictionary(5, 6, 7, 8)
    });
}

bool callback_arg_dictionary_array_n(std::shared_ptr<CallbackPassDictionaryArrayN> arg) {
    return arg->invoke(KArray {
        KOptional<MyDictionary>(),
        KOptional<MyDictionary>()
    });
}

bool callback_arg_interface_array(std::shared_ptr<CallbackPassInterfaceArray> arg) {
    return arg->invoke(KArray {
        std::shared_ptr<IMyInterface>(IMyInterface::_create()),
        std::shared_ptr<IMyInterface>(IMyInterface::_create())
    });
}

bool callback_arg_interface_array_n(std::shared_ptr<CallbackPassInterfaceArrayN> arg) {
    return arg->invoke(KArray {
        KOptional<std::shared_ptr<IMyInterface>>(),
        KOptional<std::shared_ptr<IMyInterface>>()
    });
}
